google-cloud-pub-sub-grpc/src/main/scala/org/apache/pekko/stream/connectors/googlecloud/pubsub/grpc/javadsl/GooglePubSub.scala

```diff
@@ -61,15 +62,20 @@ object GooglePubSub {
           .setStreamAckDeadlineSeconds(0)
           .build()
 
-        subscriber(mat, attr).client
+        val streamingPullResult: Source[StreamingPullResponse, NotUsed] = subscriber(mat, attr).client
           .streamingPull(
             Source
               .single(request)
               .concat(
                 Source
                   .tick(Duration.ZERO, pollInterval, subsequentRequest)
                   .mapMaterializedValue(cancellable.complete(_))))
-          .mapConcat(_.getReceivedMessagesList)
+
+        streamingPullResult
+          .mapConcat(new function.Function[StreamingPullResponse, java.util.List[ReceivedMessage]] {
```

Comment:
Can this be a SAM?

Reply:
Apologies, I'm not familiar with the term 'SAM'. Could you give me the full name?

Reply:
In Scala3, the autoconversion of the Scala function to a Pekko JAPI Function does not happen. To be frank, I'm not really sure how the Scala2 code does this conversion but whatever it is doing does not happen automatically in Scala2 - so the code fails to compile.

Reply:
> Apologies, I'm not familiar with the term 'SAM'. Could you give me the full name?

SAM = Single Abstract Method, terminology introduced in Java 1.8 which means a class with just a single method aka an anonymous function. It allows the boilerplate reduction of having to manually create a class with a method (which is whats happening here).

What @He-Pin is saying is that you can probably replace

```scala
new function.Function[StreamingPullResponse, java.util.List[ReceivedMessage]] {
override def apply(response: StreamingPullResponse): java.util.List[ReceivedMessage] =
              response.getReceivedMessagesList
          })
```

With just

```scala
(response: StreamingPullResponse) => response.getReceivedMessagesList
```

Reply:
thanks @mdedetrich - I did try that SAM format but Scala3 compiler continued to fail to compile the code

I'm afraid that Scala3 compiler doesn't always support stuff that works in Scala2 compiler.

Reply:
Yeah this is likely due to a regression introduced in Scala 3 where it does type inference slightly differently than Scala 2 especially when dealing with code defined in Java (i.e. `java.util.List`).

I am checking out the branch now, usually you can fix this with some manual type ascriptions.

Reply:
```
[error] -- [E007] Type Mismatch Error: /Users/pj.fanning/code/incubator-pekko-connectors/google-cloud-pub-sub-grpc/src/main/scala/org/apache/pekko/stream/connectors/googlecloud/pubsub/grpc/javadsl/GooglePubSub.scala:75:21 
[error] 75 |          .mapConcat((response: StreamingPullResponse) => response.getReceivedMessagesList)
[error]    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |         Found:    com.google.pubsub.v1.StreamingPullResponse =>
[error]    |           java.util.List[com.google.pubsub.v1.ReceivedMessage]
[error]    |         Required: org.apache.pekko.japi.function.Function[
[error]    |           com.google.pubsub.v1.StreamingPullResponse, ? <: Iterable[T]]
[error]    |
[error]    |         where:    T is a type variable with constraint 
[error]    |
[error]    | longer explanation available when compiling with `-explain`
```

The issue is that the SAM code is the same syntax as a Scala lambda function. Whatever is autoconverting the Scala function to a Java function when using Scala2 compiler just does not happen when Scala3 compiler is used.

Reply:
Indeed I am aware, we had similar issues when trying to put Scala 3 into Kafka. Figuring this out now

Reply:
@pjfanning This version works for all versions of Scala (including Scala 3)

```scala
.mapConcat(
  ((response: StreamingPullResponse) =>
        response.getReceivedMessagesList): pekko.japi.function.Function[StreamingPullResponse,
    java.lang.Iterable[ReceivedMessage]])
```

Issue appears to be due to the fact that these are custom function interfaces that Pekko has defined, I will report this upstream.

Reply:
Reported upstream at https://contributors.scala-lang.org/t/better-type-inference-for-scala-send-us-your-problematic-cases/2410/183. You can also add this link as a comment in the code.

Reply:
thanks - uptaken